test/list_test.go

```diff
@@ -178,32 +178,100 @@ func TestList(t *testing.T) {
 	}
 }
 
-// BenchmarkList is a benchmark for the Get operation.
 func BenchmarkList(b *testing.B) {
-	for _, backendType := range []string{endpoint.SQLiteBackend, endpoint.DQLiteBackend} {
-		b.Run(backendType, func(b *testing.B) {
-			b.StopTimer()
-			g := NewWithT(b)
+	setup := func(ctx context.Context, tx *sql.Tx, payloadSize, n int) error {
+		if err := insertMany(ctx, tx, "key", payloadSize, n); err != nil {
+			return err
+		}
+		b.Log("insert", n)
+
+		if err := updateMany(ctx, tx, "key", payloadSize, n/2); err != nil {
+			return err
+		}
+		b.Log("update", n)
+
+		if err := deleteMany(ctx, tx, "key", n/2); err != nil {
+			return err
+		}
+		b.Log("delete", n)
+		return nil
+	}
+	backends := []string{endpoint.SQLiteBackend, endpoint.DQLiteBackend}
+	for _, backendType := range backends {
+		payloads := []struct {
+			name string
+			size int
+		}{{
+			name: "tiny",
+			size: 100,
+		}, {
+			name: "page-fitting",
+			size: 1000,
+		}, {
+			name: "page-overflowing",
+			size: 5000,
+		}}
+		for _, payload := range payloads {
+			b.Run(fmt.Sprintf("%s-%s", backendType, payload.name), func(b *testing.B) {
+				b.Run("all", func(b *testing.B) {
+					b.StopTimer()
+					g := NewWithT(b)
+
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+
+					kine := newKineServer(ctx, b, &kineOptions{
+						backendType: backendType,
+						setup: func(ctx context.Context, tx *sql.Tx) error {
+							return setup(ctx, tx, payload.size, b.N)
+						},
+					})
+
+					kine.ResetMetrics()
+					b.StartTimer()
+					resp, err := kine.client.Get(ctx, "key/", clientv3.WithPrefix())
 
-			ctx, cancel := context.WithCancel(context.Background())
-			defer cancel()
+					g.Expect(err).To(BeNil())
+					g.Expect(resp.Kvs).To(HaveLen((b.N + 1) / 2))
+					kine.ReportMetrics(b)
+				})
 
-			kine := newKineServer(ctx, b, &kineOptions{
-				backendType: backendType,
-				setup: func(db *sql.DB) error {
-					return setupScenario(ctx, db, "key", b.N*2, b.N, b.N)
-				},
+				b.Run("pagination", func(b *testing.B) {
+					b.StopTimer()
+					g := NewWithT(b)
+
+					ctx, cancel := context.WithCancel(context.Background())
+					defer cancel()
+
+					kine := newKineServer(ctx, b, &kineOptions{
+						backendType: backendType,
+						setup: func(ctx context.Context, tx *sql.Tx) error {
```

Comment:
Not sure if this is something we can change in this PR, but it's not immediately clear what this is setting up for me, it looks like it's setting up a sql channel, could this be renamed? I'm not quite sure what the name here would be. Otherwise, this is okay

Reply:
The setup is not making any channel. It's setting up the scenario by bulk-inserting randomly generated rows, so that we can benchmark over a big dataset. 

Doing that with the client would make a test like this take dramatically longer (event 10-50x) and wouldn't be very useful.